vmcnet/updates/proxsr.py

```diff
@@ -0,0 +1,108 @@
+"""Stochastic reconfiguration (SR) routine."""
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+
+from vmcnet.utils.typing import Array, ModelApply, P, Tuple
+
+import chex
+from vmcnet.utils.pytree_helpers import (
+    multiply_tree_by_scalar,
+    tree_inner_product,
+)
+from vmcnet import utils
+
+
+def get_proxsr_update_fn(
+    log_psi_apply: ModelApply[P],
+    damping_type: str = "diag_shift",
+    damping: chex.Scalar = 0.001,
+    prev_grad_decay: chex.Scalar = 0.99,
+):
+    """
+    Get the ProxSR update function.
+
+    Args:
+        log_psi_apply (Callable): computes log|psi(x)|, where the signature of this
+            function is (params, x) -> log|psi(x)|
+        damping_type (str): either "diag_shift" or "pinv"
+        damping (float): damping parameter
+        prev_grad_decay (float): ProxSR-specific parameter
+
+    Returns:
+        Callable: ProxSR update function. Has the signature
+        (centered_energies, params, prev_grad, positions) -> new_grad
+    """
+
+    def raveled_log_psi_grad(params: P, positions: Array) -> Array:
+        log_grads = jax.grad(log_psi_apply)(params, positions)
+        return jax.flatten_util.ravel_pytree(log_grads)[0]
+
+    batch_raveled_log_psi_grad = jax.vmap(raveled_log_psi_grad, in_axes=(None, 0))
+
+    def invert_T(T, nchains):
+        eigval, eigvec = jnp.linalg.eigh(T)
+
+        # should be nonnegative since it's PSDF matrix
+        eigval = jnp.where(eigval < 0, 0, eigval)
+        # Damping has scale factor of nchains since we didn't divide T
```

Comment:
Very small nit: add empty line before this comment and remove empty line after :)